### PR TITLE
Fix possible CANNOT_READ_ALL_DATA during ZooKeeper client finalization and add some tests

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible `CANNOT_READ_ALL_DATA` during ZooKeeper client finalization and add some tests for `clickhouse-keeper-client`

Reverts: #54616 (cc @alexey-milovidov )
Resubmits: #54547 (cc @pufit )

### CODE REVIEW

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 1]
```diff
@@ -703,6 +703,8 @@ void ZooKeeper::receiveThread()
             {
                 if (requests_queue.isFinished())
                     break;
+                if (finalization_started.test())
+                    break;
```
Comment (tavplubix) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:707]: Why? `CANNOT_READ_ALL_DATA` means that we failed to read Keeper's response from the socket. It does not seem to be related to the finalization

Comment (azat) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp:707]: finalize firstly calls `shutdown()` on the socket, this will trigger `poll(POLLIN)` (EOF) and the `receiveThread` will trigger a `receiveEvent` which will fail because `read` will return `0` (instead of `4` - length).

Note, that rearrange the code in the finalize (i.e. call `shutdown()` after) since the client sends CLOSE_XID before, and the server may close the connection already, and this will also trigger `poll(POLLIN)` (EOF), and `read` will return `0` also.

Comment (tavplubix) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp:707]: Got it, thanks. So do we still need the `if (requests_queue.isFinished())` condition?

**Review by tavplubix — APPROVED** [review]

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp (hunk 1 above)
Comment (azat) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp:707]: Oh, you are right, could be removed. Rebased.

**Comment by azat** [comment]
>Integration tests (asan, analyzer) [6/6] — fail: 31, passed: 178

- `test_cluster_copier`

Looks like some problems with docker, though logs does not have enough details to understand the root cause:

```
time="2023-09-14T19:12:22.977295892Z" level=debug msg="Calling POST /v1.42/containers/create?name=roottestclustercopier_zoo2_1"
...
time="2023-09-14T19:17:16.301662695Z" level=debug msg="Calling GET /v1.42/containers/json?all=1&filters=%7B%22name%22%3A%7B%22%5E%2Froottestclustercopier_.%2A_1%24%22%3Atrue%7D%7D"
```

- `test_dictionaries_ddl`
- `test_dictionaries_mysql`

Can't wait for MySQL, apparently worker was too slow, from `test_dictionaries_ddl/_instances_0/mysql/logs/general.log`:

```
mysqld, Version: 5.7.43-log (MySQL Community Server (GPL)). started with:
Tcp port: 0  Unix socket: /var/run/mysqld/mysqld.sock
Time                 Id Command    Argument
2023-09-14T19:26:02.489698Z         1 Query     CREATE DATABASE mysql;
...
2023-09-14T19:29:15.393200Z         1 Query     INSERT INTO help_topic (help_topic_id,help_category_id,name,description,example,url) VALUES (268,21,'IS_FREE_LOCK','Syntax:\nIS_FREE_LOCK(str)\n\nChecks whether the lock named str is free to use (that is, not locked).\nReturns 1 if the lock is free (no one is using the lock), 0 if the lock\nis in use, and NULL if an error occurs (such as an incorrect argument).\n\nURL: https://dev.mysql.com/doc/refman/5.7/en/locking-functions.html\n\n','','https://dev.mysql.com/doc/refman/5.7/en/locking-functions.html');
```

While the timeout was 180 sec, I doubt that it worth increasing, I guess there will be lots of other places that will fail on such a slow machine.

- `test_default_compression_codec`

```
E           subprocess.TimeoutExpired: Command '['docker', 'exec', '-u', 'root', 'roottestdefaultcompressioncodec_node3_1', 'bash', '-c', 'cp /usr/share/clickhouse_fresh /usr/bin/clickhouse && chmod 777 /usr/bin/clickhouse']' timed out after 300 seconds
```

Also no details in dockerd logs:

```
time="2023-09-14T19:26:56.183410529Z" level=debug msg="form data: {\"AttachStderr\":true,\"AttachStdin\":false,\"AttachStdout\":true,\"Cmd\":[\"bash\",\"-c\",\"cp /usr/share/clickhouse_fresh /usr/bin/clickhouse \\u0026\\u0026 chmod 777 /usr/bin/clickhouse\"],\"Detach\":false,\"DetachKeys\":\"\",\"Env\":null,\"Privileged\":false,\"Tty\":false,\"User\":\"root\",\"WorkingDir\":\"\"}"
time="2023-09-14T19:26:56.183698091Z" level=debug msg="Calling POST /v1.42/exec/82af04df394a7ccb360b25c5bc2fe68dba969f201a541d8c0ff4d5de683b0860/start"
time="2023-09-14T19:26:56.183743761Z" level=debug msg="form data: {\"Detach\":false,\"Tty\":false}"
time="2023-09-14T19:26:56.183803922Z" level=debug msg="starting exec command 82af04df394a7ccb360b25c5bc2fe68dba969f201a541d8c0ff4d5de683b0860 in container 3b7060d6059cf249dccab89e355cbe52ddd745510c6d7d7f354ff7d712bede30"
```

**Comment by tavplubix** [comment]
Hm, it fails in master for some reason: https://s3.amazonaws.com/clickhouse-test-reports/0/f3294423692e27811fca71227e09e6340ca93d7f/stateless_tests__asan__[2_4].html

**Comment by azat** [comment]
Let me fix it shortly
